core/shared/src/main/scala/scalaz/zio/ZManaged.scala

```diff
@@ -73,93 +213,516 @@ final case class ZManaged[-R, +E, +A](reserve: ZIO[R, E, Reservation[R, E, A]])
+
+  /**
+   * Executes this effect, skipping the error but returning optionally the success.
+   */
+  final def option: ZManaged[R, Nothing, Option[A]] =
+    fold(_ => None, Some(_))
+
+  /**
+   * Translates effect failure into death of the fiber, making all failures unchecked and
+   * not a part of the type of the effect.
+   */
+  final def orDie(implicit ev: E <:< Throwable): ZManaged[R, Nothing, A] =
+    orDieWith(ev)
+
+  /**
+   * Keeps none of the errors, and terminates the fiber with them, using
+   * the specified function to convert the `E` into a `Throwable`.
+   */
+  final def orDieWith(f: E => Throwable): ZManaged[R, Nothing, A] =
+    mapError(f).catchAll(ZManaged.die)
+
+  /**
+   * Executes this effect and returns its value, if it succeeds, but
+   * otherwise executes the specified effect.
+   */
+  final def orElse[R1 <: R, E2, A1 >: A](that: => ZManaged[R1, E2, A1]): ZManaged[R1, E2, A1] =
+    foldM(_ => that, ZManaged.succeed)
+
+  /**
+   * Returns an effect that will produce the value of this effect, unless it
+   * fails, in which case, it will produce the value of the specified effect.
+   */
+  final def orElseEither[R1 <: R, E2, B](that: => ZManaged[R1, E2, B]): ZManaged[R1, E2, Either[A, B]] =
+    foldM(_ => that.map(Right[A, B]), a => ZManaged.succeed(Left[A, B](a)))
+
+  /**
+   * Provides the `ZManaged` effect with its required environment, which eliminates
+   * its dependency on `R`.
+   */
+  final def provide(r: R): ZManaged[Any, E, A] =
+    provideSome(_ => r)
+
+  /**
+   * Provides some of the environment required to run this effect,
+   * leaving the remainder `R0`.
+   */
+  final def provideSome[R0](f: R0 => R): ZManaged[R0, E, A] =
+    Managed(reserve.provideSome(f).map(r => Reservation(r.acquire.provideSome(f), r.release.provideSome(f))))
+
+  /**
+   * Keeps some of the errors, and terminates the fiber with the rest.
+   */
+  final def refineOrDie[E1](pf: PartialFunction[E, E1])(implicit ev: E <:< Throwable): ZManaged[R, E1, A] =
+    refineOrDie(pf)(ev)
+
+  /**
+   * Keeps some of the errors, and terminates the fiber with the rest, using
+   * the specified function to convert the `E` into a `Throwable`.
+   */
+  final def refineOrDieWith[E1](pf: PartialFunction[E, E1])(f: E => Throwable): ZManaged[R, E1, A] =
+    catchAll { e =>
+      pf.lift(e).fold[ZManaged[R, E1, A]](ZManaged.die(f(e)))(ZManaged.fail)
+    }
+
+  /**
+   * Retries with the specified retry policy.
+   * Retries are done following the failure of the original `io` (up to a fixed maximum with
+   * `once` or `recurs` for example), so that that `io.retry(Schedule.once)` means
+   * "execute `io` and in case of failure, try again once".
+   */
+  final def retry[R1 <: R, E1 >: E, S](policy: ZSchedule[R1, E1, S]): ZManaged[R1 with Clock, E1, A] =
+    ZManaged {
+      self.reserve.map { case Reservation(acquire, release) => Reservation(acquire.retry(policy), release) }
```

Comment:
This would lead to double the amount of retries in failure cases though. I think this might be quite surprising. Not retrying on reserve is as well though.

Reply:
Yes, that's true. I don't have time to think this through so it may be good to defer this to another ticket as it's a very cool addition and it's worth getting it right.

Reply:
Ok I'll remove this from here and open a ticket for discussion

Reply:
I thought about this some more. What do you think about deferring this decision to the users.
I was thinking of 3 methods { retryReservation, retryAcquisition, retryBoth } with documentation explaining the difference and the gotcha involving retryBoth

Reply:
If you're a user trying to add retrying to a `ZManaged`, how would you know which one of them to use?

Reply:
In that case you would need to go with retryBoth (maybe this should be just named retry than instead?), I think.

But I think this can be a general issue with ZManaged. I think you would have similiar problems when reasoning about interruptibility of acquisition or whether the release logic will run in case of failures or only if the acquisition was successful.

But you are right that this probably needs to be properly thought through.
BTW i think one another possible approach would be to manually thread the state of the schedule through between reservation and acquisition. This should result in sane usage for users without resulting in double retries.

Reply:
Threading the schedule state sounds promising! I suggest that we defer all the problematic combinators to separate tickets so it'll be easier to discuss. This is going to be very useful even without them and I'd love to get this in!

Reply:
Agreed on deferring the tough ones. I also think having 3 different retry methods seems problematic from a usability perspective.  

I'd argue the user does not see acquire/release and if they call `retry` it will be calling retry on the `A` produced by the Managed, not the unseen, hidden resource behind the scenes.

Now maybe a separate combinator could operate on the resource acquisition, and allow doing operations on that (kind of like how `flipWith` allows doing operations on the error channel). 

Not sure but worth thinking about and getting right..